sematic/scheduling/job_scheduler.py

```diff
@@ -0,0 +1,141 @@
+# Standard Library
+import logging
+from typing import Iterable, List, Optional, Tuple
+
+# Sematic
+from sematic.abstract_future import FutureState
+from sematic.db.models.resolution import Resolution, ResolutionStatus
+from sematic.db.models.run import Run
+from sematic.scheduling import kubernetes as k8s
+from sematic.scheduling.external_job import KUBERNETES_JOB_KIND, ExternalJob
+
+logger = logging.getLogger(__name__)
+
+
+class RunStateNotSchedulable(Exception):
+    """The run is not in a state that would allow it to be scheduled."""
+
+    pass
+
+
+def schedule_run(run: Run, resolution: Resolution) -> Run:
+    """Start a job for the run on external compute.
+
+    Parameters
+    ----------
+    run:
+        The run to schedule
+    resolution:
+        The resolution associated with the run
+    """
+    run.external_jobs = _refresh_external_jobs(run.external_jobs)
+    _validate_scheduleable(run, resolution)
+    external_jobs_list = list(run.external_jobs) + [_schedule_job(run, resolution)]
+    run.external_jobs = tuple(external_jobs_list)
+    run.future_state = FutureState.SCHEDULED
+    return run
+
+
+def update_run_status(
+    future_state: FutureState, external_jobs: List[ExternalJob]
+) -> Tuple[FutureState, Optional[str]]:
+    """Determine whether a new run state should be used based ONLY external job statuses
+
+    The external jobs themselves will have their state information refreshed before
+    determining whether the run needs its status changed.
+
+    Parameters
+    ----------
+    future_state:
+        The current state of the run
+    external_jobs:
+        The external jobs associated with the run.
+    """
+    if future_state.is_terminal():
+        return future_state, None
+    if future_state.value == FutureState.RAN.value:
+        return future_state, None
```

Comment:
because `run.future_state` is ambiguously a string or an actual FutureState. The type hint says it's a `FutureState`, and you can *set* it with a `FutureState`, but when you read it it's always a string. When doing comparisons, this leads to subtle bugs because of things like `FutureState.CREATED == "CREATED"` silently not doing what you expect. If we always do `state.value == FutureState.value`, it will work if we actually got a FutureState as we expect, but if we got a string it will generate an actual error message instead of just silently passing. In other words: defensive programming 😆

Reply:
I got bit by this discrepancy a few times while writing this PR